### PR TITLE
Add pyproject.toml

### PR DESCRIPTION
**Pull request by AnandInguva**
Context

* Directly calling setup.py for installing the Python package is deprecated. python setup.py install will throw a warning:
```
SetuptoolsDeprecationWarning: setup.py install is deprecated.
Use build and pip and other standards-based tools.
```
* The new replacement for setup.py is PEP 517 and PEP 621.
     * PEP 621: The package's metadata is now specified in pyproject.toml. The TOML file is declarative and cannot contain arbitrary code unlike setup.py.

This PR provides
1. Replace `setup.py sdist` with [build](https://pypi.org/project/build/). This will install the build dependencies in an isolated environments and then generates sdist. Setuptools itself strongly suggests using this and to avoid using `setup.py sdist` since it is treated as deprecated. You can find more at https://github.com/pypa/setuptools/blob/245d72a8aa4d47e1811425213aba2a06a0bb64fa/docs/deprecated/commands.rst#L10
   * Modify the CI to build the sdist with `python -m build` 

2. Remove the `build-requirements.txt` since we move those dependencies into `pyproject.toml`. 
3. Remove the code in `gen_protos.py` where we install `build dependencies`.  This is not required since build-requirements are always installed from pyporject.toml

Other notable changes:
* Change several Linux-based unit test suites to install SDK from a wheel. This speeds up suites consisting of multiple isolated builds by reusing the the binary distribution that is built once for entire suite.
* `grpcio` is not importable during distribution creation, therefore call `gen_protos.py` in a subprocess. 
* The format of sdist is changed from `.zip` to `.tar.gz` since we are going to use `build` module for building sdist

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by AnandInguva** [comment]
R : @tvalentyn

File: sdks/python/scripts/run_pytest.sh [hunk 1]
```diff
@@ -42,10 +42,10 @@ echo "posargs: $posargs"
 
 # Run with pytest-xdist and without.
 pytest -o junit_suite_name=${envname} \
-  --junitxml=pytest_${envname}.xml -m 'not no_xdist' -n 6 ${pytest_args} --pyargs ${posargs}
+  --junitxml=pytest_${envname}.xml -m 'not no_xdist' -n 6 --import-mode=importlib ${pytest_args} --pyargs ${posargs}
```
Comment (AnandInguva) [inline sdks/python/scripts/run_pytest.sh:45]: Newer projects needs to use importlib. Also using importlib, we are able to use installed version of Apache Beam for testing in tox environments instead of the local source(without this flag installed version gets shadowed by the local source)

https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html#which-import-mode

File: sdks/python/scripts/run_tox.sh [hunk 2]
```diff
@@ -56,12 +56,13 @@ fi
 if [[ ! -z $2 ]]; then
   SDK_LOCATION="$1"
   shift;
-  tox -c tox.ini run --recreate -e "$TOX_ENVIRONMENT" --installpkg "$SDK_LOCATION" -- "$@"
+  tox -rvv -c tox.ini run --recreate -e "$TOX_ENVIRONMENT" --installpkg "$SDK_LOCATION" -- "$@"
```
Comment (AnandInguva) [inline sdks/python/scripts/run_tox.sh:59]: Debug mode. Remove this flag before merging

**Comment by AnandInguva** [comment]
Run Seed Job

**Comment by AnandInguva** [comment]
retest this please

**Comment by AnandInguva** [comment]
Run Seed Job

**Comment by AnandInguva** [comment]
retest this please

File: sdks/python/test-suites/tox/common.gradle [hunk 3]
```diff
@@ -41,6 +41,26 @@ project.tasks.register("preCommitPy${pythonVersionSuffix}") {
       if (pythonVersionSuffix.equals('38')) {
           dependsOn = ["testPy38Cython"]
       } else {
-          dependsOn = ["testPy${pythonVersionSuffix}Cloud", "testPy${pythonVersionSuffix}Cython"]
+          dependsOn = ["testPy${pythonVersionSuffix}Cloud"]
+        }
+}
+
+getVersionsAsList('python_versions').each { it ->
```
Comment (AnandInguva) [inline sdks/python/test-suites/tox/common.gradle:48]: left over work. remove this

**Comment by AnandInguva** [comment]
Run Seed Job

**Comment by AnandInguva** [comment]
retest this please

**Comment by AnandInguva** [comment]
Run Seed Job

**Comment by AnandInguva** [comment]
retest this please

File: sdks/python/scripts/run_pytest.sh (hunk 1 above)
Comment (tvalentyn) [reply sdks/python/scripts/run_pytest.sh:45]: do we need to do anything for test files with the same name (we have quite a bit of `util_test.py`)

File: sdks/python/scripts/run_tox.sh [hunk 4]
```diff
@@ -25,11 +25,11 @@
 ###########################################################################
 # Usage check.
 if [[ $# < 1 ]]; then
-  printf "Usage: \n$> ./scripts/run_tox.sh <tox_environment> [<sdk_location> [<posargs> ...]]"
-  printf "\n\ttox_environment: [required] Tox environment to run the test in.\n"
-  printf "\n\tsdk_location: [optional] SDK tarball artifact location.\n"
-  printf "\n\tposarg: [optional] Any additional arguments will be passed as posargs to tox.\n"
-  exit 1
+ printf "Usage: \n$> ./scripts/run_tox.sh <tox_environment> [<sdk_location> [<posargs> ...]]"
```
Comment (tvalentyn) [inline sdks/python/scripts/run_tox.sh:28]: nit: let's  use consistent formatting and keep two spaces

File: sdks/python/scripts/run_tox.sh [hunk 5]
```diff
@@ -53,18 +53,32 @@ if [[ "$JENKINS_HOME" != "" ]]; then
...
-  tox -c tox.ini run --recreate -e "$TOX_ENVIRONMENT"
+  shift
+fi
+
+# If SDK_LOCATION is identified and there are still arguments left, those are posargs.
+# If SDK_LOCATION is not identified, all remaining arguments are posargs.
+
+if [[ ! -z "$SDK_LOCATION" ]]; then
+  if [[ $# -gt 0 ]]; then  # There are posargs
+    tox -rvv -c tox.ini run --recreate -e "$TOX_ENVIRONMENT" --installpkg "$SDK_LOCATION" -- "$@"
```
Comment (tvalentyn) [inline sdks/python/scripts/run_tox.sh:67]: do we need -vv here and below?

File: sdks/python/scripts/run_tox.sh [hunk 6]
```diff
@@ -53,18 +53,32 @@ if [[ "$JENKINS_HOME" != "" ]]; then
...
+  else
+    tox -rvv -c tox.ini run --recreate -e "$TOX_ENVIRONMENT" --installpkg "$SDK_LOCATION"
+  fi
+else  # No SDK_LOCATION; all arguments are posargs
+  tox -c tox.ini run --recreate -e "$TOX_ENVIRONMENT" -- "$@"
```
Comment (tvalentyn) [inline sdks/python/scripts/run_tox.sh:72]: What are the cases when we run tox without SDK location?

File: sdks/python/scripts/run_tox.sh [hunk 7]
```diff
@@ -53,18 +53,32 @@ if [[ "$JENKINS_HOME" != "" ]]; then
...
 
 exit_code=$?
 # Retry once for the specific exit code 245.
 if [[ $exit_code == 245 ]]; then
-  tox -c tox.ini run --recreate -e "$TOX_ENVIRONMENT"
+  tox -c tox.ini run --recreate -e "$TOX_ENVIRONMENT" --installpkg "$SDK_LOCATION"
   exit_code=$?
 fi
 exit $exit_code
+This code first captures TOX_ENVIRONMENT and then shifts the arguments. It then checks the number of remaining arguments to decide how to proceed. If there are no remaining arguments, it assumes only the environment was provided. If there's one remaining argument, it assumes that's the SDK location. If there are two or more arguments, it takes the next one as the SDK location and passes all the remaining ones as positional arguments to the tox command.
```
Comment (tvalentyn) [inline sdks/python/scripts/run_tox.sh:82]: misplaced content

File: sdks/python/setup.py [hunk 8]
```diff
@@ -187,7 +194,29 @@ def get_portability_package_data():
   # In order to find the tree of proto packages, the directory
   # structure must exist before the call to setuptools.find_packages()
   # executes below.
+  # while True: print(os.listdir())
   generate_protos_first()
+
+  # generate cythonize extensions only if we are building a wheel or
+  # building an extension or running in editable mode.
```
Comment (tvalentyn) [inline sdks/python/setup.py:201]: is  '-e . ' an  editable_wheel mode or a different one?

File: sdks/python/setup.py [hunk 9]
```diff
@@ -213,22 +242,9 @@ def get_portability_package_data():
...
-          'apache_beam/runners/worker/operations.py',
-          'apache_beam/transforms/cy_combiners.py',
-          'apache_beam/transforms/stats.py',
-          'apache_beam/utils/counters.py',
-          'apache_beam/utils/windowed_value.py',
-      ],
-                            language_level=3),
+      ext_modules=extensions,
       install_requires=[
+          'build>=0.9.0,<0.11.0',
```
Comment (tvalentyn) [inline sdks/python/setup.py:247]: interesting. is it common to add build as a dependency ? Also why not 1.x ?

File: sdks/python/test-suites/tox/common.gradle [hunk 10]
```diff
@@ -42,5 +42,5 @@ project.tasks.register("preCommitPy${pythonVersionSuffix}") {
           dependsOn = ["testPy38Cython"]
       } else {
           dependsOn = ["testPy${pythonVersionSuffix}Cloud", "testPy${pythonVersionSuffix}Cython"]
-      }
+        }
```
Comment (tvalentyn) [inline sdks/python/test-suites/tox/common.gradle:45]: unnecessary change

File: sdks/python/container/Dockerfile [hunk 11]
```diff
@@ -45,7 +45,7 @@ RUN  \
        && \
     rm -rf /var/lib/apt/lists/* && \
 
-    pip install --upgrade setuptools && \
+    pip install --upgrade setuptools && pip install --upgrade wheel && \
```
Comment (tvalentyn) [inline sdks/python/container/Dockerfile:48]: nit: you can combine both commands

File: sdks/python/apache_beam/runners/dataflow/internal/clients/cloudbuild/__init__.py [hunk 12]
```diff
@@ -29,5 +29,3 @@
   from apache_beam.runners.dataflow.internal.clients.cloudbuild.cloudbuild_v1_messages import *
 except ImportError:
   pass
-
```
Comment (tvalentyn) [inline sdks/python/apache_beam/runners/dataflow/internal/clients/cloudbuild/__init__.py:32]: these are autogenerated clients, why do we need these changes?

File: sdks/python/setup.py (hunk 9 above)
Comment (AnandInguva) [reply sdks/python/setup.py:247]: Since we use it in stager.py, I added it so that it won't cause a breaking change for the users. 

It seems 1.x was released recently. I will change the bounds. Thanks for noticing. I made this commit a long time back :p

**Review by AnandInguva — COMMENTED** [review]
I addressed few comments. I need to take a look on the unanswered comments. In the mean time, LMK if you have anymore questions

File: sdks/python/setup.py (hunk 9 above)
Comment (tvalentyn) [reply sdks/python/setup.py:247]: i think it should at most be an optional dependency, like snappy

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 13]
```diff
@@ -3032,30 +3032,40 @@ class BeamModulePlugin implements Plugin<Project> {
...
-            project.copy { from project.pythonSdkDeps; into copiedSrcRoot }
-
-            def copiedPyRoot = "${copiedSrcRoot}/sdks/python"
-            def distTarBall = "${pythonRootDir}/build/apache-beam.tar.gz"
-            project.exec {
-              executable 'sh'
-              args '-c', ". ${project.ext.envdir}/bin/activate && cd ${copiedPyRoot} && scripts/run_tox.sh $tox_env $distTarBall '$posargs'"
+          if (project.hasProperty('useWheelDistribution')) {
+            def pythonVersionNumber  = project.ext.pythonVersion.replace('.', '')
+            dependsOn ":sdks:python:bdistPy${pythonVersionNumber}linux"
```
Comment (tvalentyn) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:3041]: how do we know the platform for the wheel? we have many different test suites.

File: sdks/python/apache_beam/io/gcp/bigquery_test.py [hunk 14]
```diff
@@ -103,6 +102,51 @@ def _load_or_default(filename):
     return {}
 
 
+_DESTINATION_ELEMENT_PAIRS = [
```
[the remark this answers is not in the document]
Comment (tvalentyn) [reply sdks/python/apache_beam/io/gcp/bigquery_test.py:105]: will it be importable if we share a test util file?

File: sdks/python/container/py310/base_image_requirements.txt [hunk 15]
```diff
@@ -26,11 +26,12 @@ astunparse==1.6.3
 attrs==23.1.0
```
Comment (tvalentyn) [inline sdks/python/container/py310/base_image_requirements.txt:26]: i'd revert changes to these for now while you are iterating on the pr.

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy (hunk 13 above)
Comment (AnandInguva) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:3041]: All the tests that use this flag run on Linux.

File: sdks/python/gen_protos.py [hunk 16]
```diff
@@ -511,24 +472,23 @@ def generate_proto_files(force=False):
   if not os.path.exists(PYTHON_OUTPUT_PATH):
     os.mkdir(PYTHON_OUTPUT_PATH)
 
-  grpcio_install_loc = ensure_grpcio_exists()
   protoc_gen_mypy = _find_protoc_gen_mypy()
-  with PythonPath(grpcio_install_loc):
+  try:
```
[the remark this answers is not in the document]
Comment (AnandInguva) [reply sdks/python/gen_protos.py:476]: Due to the nature of virtual builds, the current structure of `apache_beam` gets copied to a `tmp` directory in a virtual env and installation happens. Here the `gen_protos.py` won't get copied to that directory. I tried specifying the gen_protos.py in packaging data of setuptools.setup() call in setup.py but it was not working. So I used the path to the `gen_protos.py` to run it similar to how numpy generates cython extensions https://github.com/numpy/numpy/blob/609c5399d0c2ca284ccc4b84d606727ed5640469/setup.py#L344

File: sdks/python/apache_beam/runners/common.py [hunk 17]
```diff
@@ -762,6 +762,7 @@ def __init__(self,
     # Try to prepare all the arguments that can just be filled in
     # without any additional work. in the process function.
     # Also cache all the placeholders needed in the process function.
+    input_args = list(input_args)
```
[the remark this answers is not in the document]
Comment (AnandInguva) [reply sdks/python/apache_beam/runners/common.py:765]: Sometimes this gets a list and sometimes this gets a tuple. Running tests using installed version of beam complaints about it and fails. So force casting this variable to list.

File: sdks/python/setup.py [hunk 18]
```diff
@@ -155,12 +156,18 @@ def cythonize(*args, **kwargs):
...
-
-  except ImportError:
-    warnings.warn("Could not import gen_protos, skipping proto generation.")
+    # Pyproject toml build happens in isolated environemnts. In those envs,
+    # gen_protos is unable to get imported. so we run a subprocess call.
+    cwd = os.path.abspath(os.path.dirname(__file__))
+    p = subprocess.call([
+      sys.executable,
+      os.path.join(cwd, 'gen_protos.py'),
+      '--no-force'
```
[the remark this answers is not in the document]
Comment (AnandInguva) [reply sdks/python/setup.py:165]: We don't want to generate stubs if they are already generated. This is the same as the old setup.

File: sdks/python/setup.py [hunk 19]
```diff
@@ -155,12 +156,18 @@ def cythonize(*args, **kwargs):
...
+    ])
+    if p != 0:
+      raise RuntimeError()
+  except RuntimeError:
+    logging.warning('Could not import gen_protos, skipping proto generation.')
```
[the remark this answers is not in the document]
Comment (AnandInguva) [reply sdks/python/setup.py:170]: Sorry. I didn't understand the question.


I raised the warning similarly to the older setup https://github.com/apache/beam/blob/0b131c9ae7cafd7a43f875b0df1fb714683bdcda/sdks/python/setup.py#L162.

File: sdks/python/scripts/run_tox.sh (hunk 6 above)
Comment (AnandInguva) [reply sdks/python/scripts/run_tox.sh:72]: Pylint, mypy, docs etc. Tox will take. care of packaging for us in that case.

File: sdks/python/apache_beam/runners/dataflow/internal/clients/cloudbuild/__init__.py (hunk 12 above)
Comment (AnandInguva) [reply sdks/python/apache_beam/runners/dataflow/internal/clients/cloudbuild/__init__.py:32]: Removed this changes. They were some errors saying `__path__ ` was not defined initially but they are solved now

**Comment by AnandInguva** [comment]
There was some confusion in the commits on my part so I had to force commit again.

**Comment by AnandInguva** [comment]
Run Seed Job

**Comment by AnandInguva** [comment]
Run Seed Job

**Comment by AnandInguva** [comment]
retest them please

**Comment by AnandInguva** [comment]
retest this please

**Comment by AnandInguva** [comment]
Run Seed Job

**Comment by AnandInguva** [comment]
retest this please

File: sdks/python/setup.py (hunk 8 above)
Comment (AnandInguva) [reply sdks/python/setup.py:200]: -e is a editable installation.  During editable installations, the pep 517 creates an editable wheel.  There is no `editable_wheel` mode.

File: .github/workflows/build_wheels.yml [hunk 20]
```diff
@@ -117,15 +112,15 @@ jobs:
           echo "RELEASE_VERSION=$RELEASE_VERSION" >> $GITHUB_OUTPUT
       - name: Build source
         working-directory: ./sdks/python
-        run: python setup.py sdist --formats=zip
+        run: pip install -U build && python -m build --sdist
       - name: Add checksums
         working-directory: ./sdks/python/dist
         run: |
-          file=$(ls | grep .zip | head -n 1)
+          file=$(ls | grep .tar.gz | head -n 1)
```
[the remark this answers is not in the document]
Comment (AnandInguva) [reply .github/workflows/build_wheels.yml:119]: Filed an issue https://github.com/apache/beam/issues/28629. Since we have time for the next release cut, I will make changes in a follow up PR.

**Comment by AnandInguva** [comment]
@tvalentyn I addressed the comments and verified tests are working as expected. PTAL

File: sdks/python/apache_beam/ml/gcp/naturallanguageml_test.py [hunk 21]
```diff
@@ -60,12 +60,15 @@ def test_document_source(self):
     self.assertFalse('content' in dict_)
     self.assertTrue('gcs_content_uri' in dict_)
 
+  @unittest.skip(
+      'TypeError: Expected bytes, got MagicMock.'
+      'Please look at https://github.com/apache/beam/issues/28307.')
```
[the remark this answers is not in the document]
Comment (AnandInguva) [reply sdks/python/apache_beam/ml/gcp/naturallanguageml_test.py:65]: `apache_beam.ml.gcp.naturallanguageml._AnnotateTextFn._get_api_client` gets  MagicMock instead of bytes. That is the error I was getting. I will investigate this but this should be non-critical to this PRl

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 22]
```diff
@@ -3032,30 +3032,40 @@ class BeamModulePlugin implements Plugin<Project> {
...
-            // to isolate them per tox suite to avoid conflict when running
-            // multiple tox suites in parallel.
-            project.copy { from project.pythonSdkDeps; into copiedSrcRoot }
-
-            def copiedPyRoot = "${copiedSrcRoot}/sdks/python"
-            def distTarBall = "${pythonRootDir}/build/apache-beam.tar.gz"
-            project.exec {
-              executable 'sh'
-              args '-c', ". ${project.ext.envdir}/bin/activate && cd ${copiedPyRoot} && scripts/run_tox.sh $tox_env $distTarBall '$posargs'"
+          if (project.hasProperty('useWheelDistribution')) {
```
[the remark this answers is not in the document]
Comment (tvalentyn) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:3039]: wdyt about automatically adding `useWheelDistribution` whenever platform is linux so that we don't have to add it in individual test suites?

File: sdks/python/gen_protos.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (tvalentyn) [reply sdks/python/gen_protos.py:476]: this file is a bit hard to review. for changes like this, try to make separate commits to make reviewing easier: separate formatting changes from code changes, capture rationale in commit metadata.

**Review by tvalentyn — CHANGES_REQUESTED** [review]

**Comment by AnandInguva** [comment]
Run Seed Job

**Comment by AnandInguva** [comment]
retest this please

**Comment by AnandInguva** [comment]
https://github.com/apache/beam/pull/28385#discussion_r1340561540

Some tasks like mypy, pylint doesn't need installation from wheels and they use `toxTask` and they run on linux. I think it would be good idea to go the current way

**Comment by AnandInguva** [comment]
R: @tvalentyn

File: sdks/python/gen_protos.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (tvalentyn) [reply sdks/python/gen_protos.py:476]: updated the decription.

File: sdks/python/apache_beam/runners/portability/stager.py [hunk 23]
```diff
@@ -771,13 +772,26 @@ def _build_setup_package(setup_file,  # type: str
...
-        build_setup_args = [
-            Stager._get_python_executable(),
-            os.path.basename(setup_file),
-            'sdist',
-            '--dist-dir',
-            temp_dir
-        ]
+        # if build is installed in the user env, use it to
+        # build the sdist else fallback to legacy setup.py sdist call.
+        if importlib.util.find_spec('build'):
```
Comment (tvalentyn) [inline sdks/python/apache_beam/runners/portability/stager.py:777]: Are you aware of any cases where the new process might fail to build but old one would succeed? I suppose users would have a workaround to uninstall `build`.

File: sdks/python/tox.ini [hunk 24]
```diff
@@ -17,7 +17,7 @@
 
 [tox]
 # new environments will be excluded by default unless explicitly added to envlist.
-envlist = py38,py39,py310,py311,py38-{cloud,cython,docs,lint,mypy,cloudcoverage,dask},py39-{cloud,cython},py310-{cloud,cython,dask},py311-{cloud,cython,dask},whitespacelint
```
Comment (tvalentyn) [inline sdks/python/tox.ini:20]: this actually removes a test suite that covers `no cloud dependencies` variant.

A while ago we had plain suite, gcp suite , and cython suite. Later we said, why don't we get rid of one set of tests by having: `cython` + `no cloud` as one suite, and cloud as another suite. 

Loss of coverage would impact the scenario when one installs:

pip install apache-beam (without cloud/gcp extras). One could argue - should we even have this test suite or codepath? but that would be a separate conversation. For now, we could perhaps rename the `cython` suite into `noextras`.

File: sdks/python/apache_beam/io/gcp/shared_test_variables.py [hunk 25]
```diff
@@ -0,0 +1,63 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (tvalentyn) [inline sdks/python/apache_beam/io/gcp/shared_test_variables.py:2]: This file looks a bit awkward here.. also looking closer, the stub data in  `bigquery_file_loads_test`  is now also split across two files, which looks awkward too. A file like this would be better placed in /tests subdir in this folder, and could include all the necessary shared BQ stubs data & schema in one place.

Alternatively, we can reverse the course of my previous suggestion and redefine `_ELEMENTS` in `sdks/python/apache_beam/io/gcp/bigquery_test.py`. Since originally we have `_ELEMENTS = [elm[1] for elm in _DESTINATION_ELEMENT_PAIRS]` to avoid extra copypasting, it can be  redefined here as sth like: `[{'name': 'beam', 'language': 'py'}, {'name': 'beam', 'language': 'java'}, {'name': 'beam', 'language': 'go'}, {'name': 'flink', 'language': 'java'}, {'name': 'flink', 'language': 'scala'}, {'name': 'spark', 'language': 'scala'}, {'name': 'spark', 'language': 'py'}, {'name': 'spark', 'language': 'scala'}, {'name': 'beam', 'foundation': 'apache'}, {'name': 'flink', 'foundation': 'apache'}, {'name': 'spark', 'foundation': 'apache'}]`

**Review by tvalentyn — CHANGES_REQUESTED** [review]
Thanks! Also PTAL at lint failures.

File: sdks/python/tox.ini (hunk 24 above)
Comment (AnandInguva) [reply sdks/python/tox.ini:20]: We already have a test suite https://github.com/apache/beam/blob/a7bf59436a664dca932c5001b750514dab2fa296/sdks/python/tox.ini#L72 that can be treated as no-extras. I will reuse it.

File: sdks/python/scripts/run_tox.sh [hunk 26]
```diff
@@ -53,18 +53,27 @@ if [[ "$JENKINS_HOME" != "" ]]; then
...
+  fi
+else  # No SDK_LOCATION; all arguments are posargs
+  tox -c tox.ini run --recreate -e "$TOX_ENVIRONMENT" -- "$@"
 fi
 
 exit_code=$?
 # Retry once for the specific exit code 245.
 if [[ $exit_code == 245 ]]; then
-  tox -c tox.ini run --recreate -e "$TOX_ENVIRONMENT"
+  tox -c tox.ini run --recreate -e "$TOX_ENVIRONMENT" --installpkg "$SDK_LOCATION"
```
[the remark this answers is not in the document]
Comment (AnandInguva) [reply sdks/python/scripts/run_tox.sh:76]: I think we should revert back to the original logic. This was unintended change.

**Comment by AnandInguva** [comment]
I addressed the comments. PTAL @tvalentyn. Thanks

File: sdks/python/apache_beam/runners/portability/stager.py (hunk 23 above)
Comment (AnandInguva) [reply sdks/python/apache_beam/runners/portability/stager.py:777]: The new process calls the old process under the hood in a way.

**Comment by AnandInguva** [comment]
Run Python 3.8 PostCommit

**Comment by AnandInguva** [comment]
Run Seed Job

File: sdks/python/scripts/run_tox.sh (hunk 26 above)
[the remark this answers is not in the document]
Comment (tvalentyn) [reply sdks/python/scripts/run_tox.sh]: shouldn't we recreate with the same params we tried to create it in the first place?

File: sdks/python/setup.py [hunk 27]
```diff
@@ -159,15 +159,14 @@ def generate_protos_first():
...
+    p = subprocess.run([
       sys.executable,
       os.path.join(cwd, 'gen_protos.py'),
       '--no-force'
-    ])
-    if p != 0:
-      raise RuntimeError()
-  except RuntimeError:
-    logging.warning('Could not import gen_protos, skipping proto generation.')
+    ], capture_output=True, check=True)
```
Comment (tvalentyn) [inline sdks/python/setup.py:165]: note that `capture_output=True,` will hide all of stdout, and in case of error, only stderr is printed.  is this intentional?

**Review by tvalentyn — APPROVED** [review]

**Comment by AnandInguva** [comment]
Run Seed Job

**Comment by AnandInguva** [comment]
retest this please

File: sdks/python/scripts/run_tox.sh (hunk 26 above)
[the remark this answers is not in the document]
Comment (AnandInguva) [reply sdks/python/scripts/run_tox.sh]: 245 is -11 which is a segmentation fault...could be caused by pytest. In that case, we should retry it but let me make this change in a different PR.

File: sdks/python/setup.py (hunk 27 above)
Comment (AnandInguva) [reply sdks/python/setup.py:165]: printing output now

**Comment by AnandInguva** [comment]
Run Seed Job

**Comment by AnandInguva** [comment]
Run Python_Coverage PreCommit

**Comment by AnandInguva** [comment]
Run Python_PVR_Flink PreCommit

**Comment by AnandInguva** [comment]
Run Java PreCommit

**Comment by AnandInguva** [comment]
https://ci-beam.apache.org/job/beam_PreCommit_Python_Coverage_Phrase/69/ - python coverage test suite passes with expected changes on Jenkins. For GHA, it will be reflected once it is merged.

**Comment by AnandInguva** [comment]
If there are any failures, please comment here and I will try to forward fix it.

**Comment by AnandInguva** [comment]
From this run - https://github.com/apache/beam/actions/runs/6502267792/job/17661040535, I see one of the tests in the postcommits is failing. It was also seen on Jenkins. I have merged the solution 20 hours back. PTAL https://github.com/apache/beam/issues/28990. That would solve the failing test. let's monitor the tests and see if they go green for the next runs.

image

**Comment by AnandInguva** [comment]
I started a run - https://github.com/apache/beam/actions/runs/6550155250 on the master. Will monitor it for any related failures to this PR.
